File: web/src/lib/officialmap.ts
// NEAK's own FIN → provider link (data/officialmap.json), held up against
// the one the cross-check works out from addresses and names. Neither
// replaces the other: the site keeps publishing its own chain, and this
// says how often the two agree — and where they do not.
import { useEffect, useState } from 'react';
import { t } from './i18n';
import type { ColDef, Row } from './eesztTable';

export type Verdict = 'agree' | 'differ' | 'onlyOfficial' | 'onlyOurs' | 'neither';

export interface OfficialRow {
  fin: string;
  group: string;
  settlement: string;
  county: string;
  neakCode: string;
  licenceSource: string;
  ourProviderIds: string[];
  ourProvider: string;
  ourLicenceIds: string[];
  ourUnits: string[];
  officialProviderIds: string[];
  officialProvider: string;
  officialTax: string;
  officialUnits: string[];
  officialUnitHasLicence: boolean;
  verdict: Verdict;
}

export interface OfficialRaw {
  schemaVersion: number;
  asOf: string;
  licenceAsOf: string;
  dataMonth: string;
  verdicts: Verdict[];
  stats: {
    praxes: number;
    withOfficial: number;
    decidable: number;
    agreement: number;
    rescuable: number;
    bySource: Record<string, Record<string, number>>;
  } & Record<Verdict, number>;
  counties: ({ county: string; praxes: number; agreement: number } & Record<Verdict, number>)[];
  rows: OfficialRow[];
}

export const VERDICT_COLORS: Record<Verdict, string> = {
  agree: '#4fd6c2',
  differ: '#ff7a59',
  onlyOfficial: '#ffb454',
  onlyOurs: '#b8b0f5',
  neither: '#8899ad',
};

let cache: OfficialRaw | null = null;
let pending: Promise<OfficialRaw | null> | null = null;

export function useOfficialMap(): OfficialRaw | null {
  const [data, setData] = useState<OfficialRaw | null>(cache);
  useEffect(() => {
    let alive = true;
    pending ??= fetch(`${import.meta.env.BASE_URL}data/officialmap.json`)
      .then((r) => (r.ok ? r.json() : null))
      .then((d: OfficialRaw | null) => { cache = d; return d; })
      .catch(() => null);
    void pending.then((d) => { if (alive) setData(d); });
    return () => { alive = false; };
  }, []);
  return data;
}

export const OFFICIAL_COLUMNS: ColDef[] = [
  { key: 'fin', labelKey: 'eeszt.colFin', type: 'text', visible: true },
  { key: 'settlement', labelKey: 'stats.thSettlement', type: 'text', visible: true },
  { key: 'county', labelKey: 'stats.thCounty', type: 'enum', visible: true },
  { key: 'verdictLabel', labelKey: 'officialmap.colVerdict', type: 'enum', visible: true },
  { key: 'sourceLabel', labelKey: 'officialmap.colSource', type: 'enum', visible: true },
  { key: 'ourProvider', labelKey: 'officialmap.colOurs', type: 'text', visible: true },
  { key: 'officialProvider', labelKey: 'officialmap.colOfficial', type: 'text', visible: true },
  { key: 'ourProviderId', labelKey: 'officialmap.colOurId', type: 'text', visible: false },
  { key: 'officialProviderId', labelKey: 'officialmap.colOfficialId', type: 'text', visible: false },
  { key: 'ourLicence', labelKey: 'eeszt.colLicence', type: 'text', visible: false },
  { key: 'officialTax', labelKey: 'operating.colTax', type: 'text', visible: false },
  { key: 'groupLabel', labelKey: 'operating.colGroup', type: 'enum', visible: false },
];

export const OFFICIAL_COUNTY_COLUMNS: ColDef[] = [
  { key: 'county', labelKey: 'stats.thCounty', type: 'enum', visible: true },
  { key: 'praxes', labelKey: 'officialmap.colPraxes', type: 'number', visible: true },
  { key: 'agree', labelKey: 'officialmap.verdict.agree', type: 'number', visible: true },
  { key: 'differ', labelKey: 'officialmap.verdict.differ', type: 'number', visible: true },
  { key: 'agreementPct', labelKey: 'officialmap.colAgreement', type: 'number', visible: true },
];

export function officialRows(data: OfficialRaw | null, only?: Verdict): Row[] {
  return (data?.rows ?? [])
    .filter((r) => !only || r.verdict === only)
    .map((r) => ({
      fin: r.fin,
      settlement: r.settlement,
      county: r.county,
      verdictLabel: t(`officialmap.verdict.${r.verdict}`),
      sourceLabel: t(`officialmap.source.${r.licenceSource}`),
      ourProvider: r.ourProvider,
      officialProvider: r.officialProvider,
      ourProviderId: r.ourProviderIds.join(', '),
      officialProviderId: r.officialProviderIds.join(', '),
      ourLicence: r.ourLicenceIds.join(', '),
      officialTax: r.officialTax,
      groupLabel: r.group,
    }));
}

export function officialCountyRows(data: OfficialRaw | null): Row[] {
  return (data?.counties ?? []).map((c) => ({
    county: c.county,
    praxes: c.praxes,
    agree: c.agree,
    differ: c.differ,
    agreementPct: Math.round(c.agreement * 1000) / 10,
  }));
}

File: etl/officialmap.py
"""NEAK's own FIN → provider link, held up against ours.

The cross-check (crosscheck.py) works out which operating licence a praxis
runs under by matching addresses, tax numbers and provider names, because
the register we had did not say. The extended financing register does say:
NEAK_FINSZOLG_EXT carries the same rows as NEAK_FINSZOLG plus the provider's
six-character id (NNGYK6), name and tax number — the very link the
cross-check reconstructs.

This module does not replace our answer with theirs. It publishes both, side
by side, with a verdict per praxis:

    agree         the licence we publish belongs to the provider NEAK names
    differ        it belongs to someone else
    onlyOfficial  we found no licence, NEAK names a provider
    onlyOurs      we found a licence, NEAK names no provider
    neither       nothing on either side

That is the honest shape for a second opinion. Where the two disagree, the
disagreement is the finding — not a correction to be applied quietly. The
site keeps showing its own chain, and this table says how far it can be
trusted.

Usage:
  python etl/officialmap.py
"""
from __future__ import annotations

import collections
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from build_eeszt import EesztError, latest_date, load, resolve

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "data" / "officialmap.json"
SCHEMA_VERSION = 1

VERDICTS = ("agree", "differ", "onlyOfficial", "onlyOurs", "neither")


def build() -> dict:
    date = latest_date()
    ex, ext = load("neak_finszolg_ext", date)
    lx, licences = load("euszolg_engedely", date)

    operating_path = ROOT / "data" / "operating.json"
    if not operating_path.exists():
        raise EesztError("run etl/operating.py first — there is nothing to compare")
    operating = json.loads(operating_path.read_text(encoding="utf-8"))

    # which provider owns each organisational unit, and which units have a licence
    provider_of_unit: dict[str, str] = {}
    for r in licences:
        provider_of_unit.setdefault(r[lx["SZERVEZETI_EGYSEG_KOD"]],
                                    r[lx["EUSZOLG_AZONOSITO"]])

    official: dict[str, list[dict]] = collections.defaultdict(list)
    for r in ext:
        official[r[ex["FINKOD"]]].append({
            "providerId": r[ex["NNGYK6"]] or "",
            "providerName": r[ex["NNGYK_NEV"]] or "",
            "tax": r[ex["NNGYK_ADOSZAM"]] or "",
            "unit": r[ex["NNGYK9_KOD"]] or "",
            "type": r[ex["TIP"]] or "",
        })

    rows: list[dict] = []
    for praxis in operating["rows"]:
        theirs = official.get(praxis["fin"], [])
        their_ids = sorted({t["providerId"] for t in theirs if t["providerId"]})
        our_ids = sorted({provider_of_unit.get(lic["unit"], "")
                          for lic in praxis["licences"]} - {""})
        rows.append({
            "fin": praxis["fin"],
            "group": praxis["group"],
            "settlement": praxis["settlement"],
            "county": praxis["county"],
            "neakCode": praxis["neakCode"],
            "licenceSource": praxis["licenceSource"],
            "ourProviderIds": our_ids,
            "ourProvider": praxis["provider"],
            "ourLicenceIds": sorted({lic["licenceId"] for lic in praxis["licences"]
                                     if lic["licenceId"]}),
            "ourUnits": praxis["units"],
            "officialProviderIds": their_ids,
            "officialProvider": next((t["providerName"] for t in theirs
                                      if t["providerName"]), ""),
            "officialTax": next((t["tax"] for t in theirs if t["tax"]), ""),
            "officialUnits": sorted({t["unit"] for t in theirs if t["unit"]}),
            # can the official unit be followed into the licence register?
            "officialUnitHasLicence": any(t["unit"] in provider_of_unit for t in theirs),
            "verdict": verdict_of(our_ids, their_ids),
        })

    out = {
        "schemaVersion": SCHEMA_VERSION,
        "asOf": resolve("neak_finszolg_ext", date),
        "licenceAsOf": resolve("euszolg_engedely", date),
        "dataMonth": operating.get("dataMonth", ""),
        "verdicts": list(VERDICTS),
        "stats": overall(rows),
        "counties": by_county(rows),
        "rows": rows,
    }
    guard(out)
    return out


def verdict_of(ours: list[str], theirs: list[str]) -> str:
    if ours and theirs:
        return "agree" if set(ours) & set(theirs) else "differ"
    if theirs:
        return "onlyOfficial"
    if ours:
        return "onlyOurs"
    return "neither"


def overall(rows: list[dict]) -> dict:
    counts = collections.Counter(r["verdict"] for r in rows)
    by_source: dict[str, dict] = {}
    for source in ("code", "crosscheck", "none"):
        group = [r for r in rows if r["licenceSource"] == source]
        inner = collections.Counter(r["verdict"] for r in group)
        by_source[source] = {"praxes": len(group),
                             **{v: inner.get(v, 0) for v in VERDICTS}}
    decidable = counts["agree"] + counts["differ"]
    return {
        "praxes": len(rows),
        "withOfficial": sum(1 for r in rows if r["officialProviderIds"]),
        **{v: counts.get(v, 0) for v in VERDICTS},
        "decidable": decidable,
        "agreement": counts["agree"] / decidable if decidable else 0.0,
        "bySource": by_source,
        # where our chain found nothing, does the official one open a door?
        "rescuable": sum(1 for r in rows
                         if r["verdict"] == "onlyOfficial" and r["officialUnitHasLicence"]),
    }


def by_county(rows: list[dict]) -> list[dict]:
    groups: dict[str, list[dict]] = collections.defaultdict(list)
    for r in rows:
        groups[r["county"]].append(r)
    out = []
    for county, list_ in sorted(groups.items()):
        counts = collections.Counter(r["verdict"] for r in list_)
        decidable = counts["agree"] + counts["differ"]
        out.append({
            "county": county,
            "praxes": len(list_),
            **{v: counts.get(v, 0) for v in VERDICTS},
            "agreement": counts["agree"] / decidable if decidable else 0.0,
        })
    return sorted(out, key=lambda c: c["agreement"])


def guard(out: dict) -> None:
    rows = out["rows"]
    st = out["stats"]
    if len(rows) < 5000:
        raise EesztError(f"only {len(rows)} praxes compared")
    if sum(st[v] for v in VERDICTS) != len(rows):
        raise EesztError("the verdicts do not add up to the praxis count")
    if not 0.0 <= st["agreement"] <= 1.0:
        raise EesztError("agreement out of range")
    for source, block in st["bySource"].items():
        if sum(block[v] for v in VERDICTS) != block["praxes"]:
            raise EesztError(f"{source}: the verdicts do not add up")
    seen: set[str] = set()
    for r in rows:
        if r["verdict"] not in VERDICTS:
            raise EesztError(f"{r['fin']}: unknown verdict")
        if r["fin"] in seen:
            raise EesztError(f"duplicate praxis {r['fin']}")
        seen.add(r["fin"])
        if r["verdict"] == "agree" and not (
                set(r["ourProviderIds"]) & set(r["officialProviderIds"])):
            raise EesztError(f"{r['fin']}: agree without a shared provider")
    if sum(c["praxes"] for c in out["counties"]) != len(rows):
        raise EesztError("county counts do not add up")


def main() -> None:
    out = build()
    OUT.write_text(json.dumps(out, ensure_ascii=False, separators=(",", ":")) + "\n",
                   encoding="utf-8")
    st = out["stats"]
    print(f"wrote {OUT}: {st['praxes']} praxes compared "
          f"(official register {out['asOf']})")
    print(f"  agree {st['agree']}, differ {st['differ']}, "
          f"only official {st['onlyOfficial']}, only ours {st['onlyOurs']}, "
          f"neither {st['neither']} — agreement {st['agreement']:.1%}")
    for source, block in st["bySource"].items():
        if block["praxes"]:
            print(f"    licence from {source}: {block['praxes']} praxes, "
                  f"{block['agree']} agree, {block['differ']} differ")


if __name__ == "__main__":
    main()
